Link errors with different configuration of `theora`, `ogg`, `vorbis`, `opus` modules

<!-- Please search existing issues for potential duplicates before filing yours:	
https://github.com/godotengine/godot/issues?q=is%3Aissue	
-->

**Godot version:**
4.0

**OS/device including version:**
Windows 10

**Issue description:**

# Explicit theora enabled
```
scons module_theora_enabled=yes
```
```
[ 97%] Linking Program        ==> bin\godot.windows.tools.64.exe
   Creating library bin\godot.windows.tools.64.lib and object bin\godot.windows.tools.64.exp
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_sync_init referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_sync_clear referenced in function "protected: void __cdecl VideoStreamPlaybackTheora::clear(void)" (?clear@VideoStreamPlaybackTheora@@IEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_sync_buffer referenced in function "private: int __cdecl VideoStreamPlaybackTheora::buffer_data(void)" (?buffer_data@VideoStreamPlaybackTheora@@AEAAHXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_sync_wrote referenced in function "private: int __cdecl VideoStreamPlaybackTheora::buffer_data(void)" (?buffer_data@VideoStreamPlaybackTheora@@AEAAHXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_sync_pageout referenced in function "public: void __cdecl 
VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_stream_pagein referenced in function "private: int __cdecl VideoStreamPlaybackTheora::queue_page(struct ogg_page *)" (?queue_page@VideoStreamPlaybackTheora@@AEAAHPEAUogg_page@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_stream_packetout referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_stream_init referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_stream_clear referenced in function "protected: void __cdecl VideoStreamPlaybackTheora::clear(void)" (?clear@VideoStreamPlaybackTheora@@IEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_page_bos referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol ogg_page_serialno referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_info_init referenced in function "public: void __cdecl 
VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_info_clear referenced in function "protected: void __cdecl VideoStreamPlaybackTheora::clear(void)" (?clear@VideoStreamPlaybackTheora@@IEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_comment_init referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_comment_clear referenced in function "protected: void __cdecl VideoStreamPlaybackTheora::clear(void)" (?clear@VideoStreamPlaybackTheora@@IEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_block_init referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_block_clear referenced in function "protected: void __cdecl VideoStreamPlaybackTheora::clear(void)" (?clear@VideoStreamPlaybackTheora@@IEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_dsp_clear referenced in function "protected: void __cdecl VideoStreamPlaybackTheora::clear(void)" (?clear@VideoStreamPlaybackTheora@@IEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_headerin referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_init referenced in function "public: void __cdecl VideoStreamPlaybackTheora::set_file(class String const &)" (?set_file@VideoStreamPlaybackTheora@@QEAAXAEBVString@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis referenced in function "public: virtual void 
__cdecl VideoStreamPlaybackTheora::update(float)" (?update@VideoStreamPlaybackTheora@@UEAAXM@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_blockin referenced in function "public: virtual void __cdecl VideoStreamPlaybackTheora::update(float)" (?update@VideoStreamPlaybackTheora@@UEAAXM@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_pcmout referenced in function "public: virtual void __cdecl VideoStreamPlaybackTheora::update(float)" (?update@VideoStreamPlaybackTheora@@UEAAXM@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_read referenced in function "public: virtual 
void __cdecl VideoStreamPlaybackTheora::update(float)" (?update@VideoStreamPlaybackTheora@@UEAAXM@Z)
bin\godot.windows.tools.64.exe : fatal error LNK1120: 24 unresolved externals
scons: *** [bin\godot.windows.tools.64.exe] Error 1120
scons: building terminated because of errors.
```

# Vorbis disabled

```
scons module_stb_vorbis_enabled=no module_stb_vorbis_enabled=no module_webm_enabled=yes module_opus_enabled=yes
```
```
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_info_init referenced in function "private: bool __cdecl OpusVorbisDecoder::openVorbis(class WebMDemuxer const &)" (?openVorbis@OpusVorbisDecoder@@AEAA_NAEBVWebMDemuxer@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_info_init
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_info_clear referenced in function "private: void __cdecl OpusVorbisDecoder::close(void)" (?close@OpusVorbisDecoder@@AEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_info_clear
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_comment_init referenced in function "private: bool __cdecl 
OpusVorbisDecoder::openVorbis(class WebMDemuxer const &)" (?openVorbis@OpusVorbisDecoder@@AEAA_NAEBVWebMDemuxer@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_comment_init
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_comment_clear referenced in function "private: bool __cdecl OpusVorbisDecoder::openVorbis(class WebMDemuxer const &)" (?openVorbis@OpusVorbisDecoder@@AEAA_NAEBVWebMDemuxer@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_comment_clear
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_block_init referenced in function "private: bool __cdecl OpusVorbisDecoder::openVorbis(class WebMDemuxer const &)" (?openVorbis@OpusVorbisDecoder@@AEAA_NAEBVWebMDemuxer@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_block_init
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_block_clear referenced in function "private: void __cdecl OpusVorbisDecoder::close(void)" (?close@OpusVorbisDecoder@@AEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_block_clear
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_dsp_clear referenced in function "private: void __cdecl OpusVorbisDecoder::close(void)" (?close@OpusVorbisDecoder@@AEAAXXZ)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_dsp_clear
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_headerin referenced in function "private: bool __cdecl OpusVorbisDecoder::openVorbis(class WebMDemuxer const &)" (?openVorbis@OpusVorbisDecoder@@AEAA_NAEBVWebMDemuxer@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_synthesis_headerin
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_init referenced in function "private: bool __cdecl OpusVorbisDecoder::openVorbis(class WebMDemuxer const &)" (?openVorbis@OpusVorbisDecoder@@AEAA_NAEBVWebMDemuxer@@@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_synthesis_init
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis referenced in function "public: bool __cdecl OpusVorbisDecoder::getPCMS16(class WebMFrame &,short *,int &)" (?getPCMS16@OpusVorbisDecoder@@QEAA_NAEAVWebMFrame@@PEAFAEAH@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_synthesis
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_blockin referenced in function "public: bool __cdecl OpusVorbisDecoder::getPCMS16(class WebMFrame &,short *,int &)" (?getPCMS16@OpusVorbisDecoder@@QEAA_NAEAVWebMFrame@@PEAFAEAH@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_synthesis_blockin
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_pcmout referenced in function "public: bool __cdecl OpusVorbisDecoder::getPCMS16(class WebMFrame &,short *,int &)" (?getPCMS16@OpusVorbisDecoder@@QEAA_NAEAVWebMFrame@@PEAFAEAH@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_synthesis_pcmout
module_webm.windows.tools.64.lib(OpusVorbisDecoder.windows.tools.64.obj) : error LNK2019: unresolved external symbol vorbis_synthesis_read referenced in function "public: bool __cdecl OpusVorbisDecoder::getPCMS16(class WebMFrame &,short *,int &)" (?getPCMS16@OpusVorbisDecoder@@QEAA_NAEAVWebMFrame@@PEAFAEAH@Z)
module_theora.windows.tools.64.lib(video_stream_theora.windows.tools.64.obj) : error LNK2001: unresolved external symbol vorbis_synthesis_read
bin\godot.windows.tools.64.exe : fatal error LNK1120: 13 unresolved externals
scons: *** [bin\godot.windows.tools.64.exe] Error 1120
```

Should likely prevent compiling under other module configurations with explicit dependencies (`ogg`, `stb_vorbis` etc).

--- a/SConstruct
+++ b/SConstruct
@@ -86,6 +86,7 @@ env_base.__class__.add_library = methods.add_library
 env_base.__class__.add_program = methods.add_program
 env_base.__class__.CommandNoCache = methods.CommandNoCache
 env_base.__class__.disable_warnings = methods.disable_warnings
+env_base.__class__.module_check_dependencies = methods.module_check_dependencies
 
 env_base["x86_libtheora_opt_gcc"] = False
 env_base["x86_libtheora_opt_vc"] = False
@@ -607,14 +608,12 @@ if selected_platform in platform_list:
         env.Append(CPPDEFINES=["MINIZIP_ENABLED"])
 
     editor_module_list = ["regex"]
-    for x in editor_module_list:
-        if not env["module_" + x + "_enabled"]:
-            if env["tools"]:
-                print(
-                    "Build option 'module_" + x + "_enabled=no' cannot be used with 'tools=yes' (editor), "
-                    "only with 'tools=no' (export template)."
-                )
-                Exit(255)
+    if env["tools"] and not env.module_check_dependencies("tools", editor_module_list):
+        print(
+            "Build option 'module_" + x + "_enabled=no' cannot be used with 'tools=yes' (editor), "
+            "only with 'tools=no' (export template)."
+        )
+        Exit(255)
 
     if not env["verbose"]:
         methods.no_verbose(sys, env)

--- a/methods.py
+++ b/methods.py
@@ -231,6 +231,30 @@ def disable_module(self):
     self.disabled_modules.append(self.current_module)
 
 
+def module_check_dependencies(self, module, dependencies):
+    """
+    Checks if module dependencies are enabled for a given module,
+    and prints a warning if they aren't.
+    Meant to be used in module `can_build` methods.
+    Returns a boolean (True if dependencies are satisfied).
+    """
+    missing_deps = []
+    for dep in dependencies:
+        opt = "module_{}_enabled".format(dep)
+        if not opt in self or not self[opt]:
+            missing_deps.append(dep)
+
+    if missing_deps != []:
+        print(
+            "Disabling '{}' module as the following dependencies are not satisfied: {}".format(
+                module, ", ".join(missing_deps)
+            )
+        )
+        return False
+    else:
+        return True
+
+
 def use_windows_spawn_fix(self, platform=None):
 
     if os.name != "nt":

--- a/modules/opus/config.py
+++ b/modules/opus/config.py
@@ -1,5 +1,5 @@
 def can_build(env, platform):
-    return True
+    return env.module_check_dependencies("opus", ["ogg"])
 
 
 def configure(env):

--- a/modules/theora/config.py
+++ b/modules/theora/config.py
@@ -1,5 +1,5 @@
 def can_build(env, platform):
-    return True
+    return env.module_check_dependencies("theora", ["ogg", "vorbis"])
 
 
 def configure(env):

--- a/modules/vorbis/config.py
+++ b/modules/vorbis/config.py
@@ -1,5 +1,5 @@
 def can_build(env, platform):
-    return True
+    return env.module_check_dependencies("vorbis", ["ogg"])
 
 
 def configure(env):

--- a/modules/webm/config.py
+++ b/modules/webm/config.py
@@ -1,5 +1,8 @@
 def can_build(env, platform):
-    return platform not in ["iphone"]
+    if platform in ["iphone"]:
+        return False
+
+    return env.module_check_dependencies("webm", ["ogg", "opus", "vorbis"])
 
 
 def configure(env):

